### PR TITLE
fix elasticsearch-5.x-plugin when use es6.x TransportClient error

### PR DESCRIPTION
**Pull request by Indifer**
**fix elasticsearch-5.x-plugin when use es6.x TransportClient error:** 
_class[class org.elasticsearch.client.transport.TransportClientNodesService] after method[addTransportAddresses] intercept failure
java.lang.IncompatibleClassChangeError: Found class org.elasticsearch.common.transport.TransportAddress, but interface was expected_

image

because the org.elasticsearch.common.transport.TransportAddress is class in es6.x

image

so, up es5.x plugin, make it inoperative on es6.x client
image

**up es6.x plugin to support TransportClient**

### CODE REVIEW

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    private void assertGetSpan(AbstractTracingSpan getSpan, Object ret) {
+        assertThat(getSpan instanceof ExitSpan, is(true));
+
+        ExitSpan span = (ExitSpan) getSpan;
+        assertThat(span.getOperationName().split("[$$]")[0], is("Elasticsearch/GetRequest"));
+        assertThat(SpanHelper.getComponentId(span), is(TRANSPORT_CLIENT.getId()));
+
+        List<TagValuePair> tags = SpanHelper.getTags(span);
+        Assert.assertTrue(tags.size() > 4);
+
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: It is recommended that the content of tags here be tested and verified.

File: apm-sniffer/apm-sdk-plugin/elasticsearch-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v5/define/GenericActionInstrumentation.java [hunk 2]
```diff
@@ -60,4 +60,10 @@ public String getConstructorInterceptor() {
     protected ClassMatch enhanceClass() {
         return byHierarchyMatch(new String[] {"org.elasticsearch.action.GenericAction"});
     }
+
+    @Override
+    protected String[] witnessClasses() {
+        return new String[]{"org.elasticsearch.common.transport.InetSocketTransportAddress"};
```
[the remark this answers is not in the document]
Comment (Indifer) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-5.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v5/define/GenericActionInstrumentation.java]: ok

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/TransportAddressCache.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        transportAddresses = nodesBuilder;
+        transportAddressesStr = format();
+    }
+
+    private String format() {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (TransportAddress node : transportAddresses) {
+            stringBuilder.append(node.getAddress()).append(":").append(node.getPort()).append(";");
```
[the remark this answers is not in the document]
Comment (Indifer) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/TransportAddressCache.java]: ok

**Comment by Indifer** [comment]
> These two plugin tests should be enhanced to test your scenarios.
> 
> 1. https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/elasticsearch-5.x-scenario
> 2. https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/elasticsearch-6.x-scenario
> 
> If you haven't familiar with the plugin test, read this doc, https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md. You should only need to use the ES APIs in the way used to have issue, and change the expected data file to prove it is working now.

ok

**Comment by wu-sheng** [comment]
@Indifer  Have you finished the change?

**Comment by wu-sheng** [comment]
@Indifer We are still waiting for your update.

**Comment by Indifer** [comment]
@wu-sheng

**Comment by Indifer** [comment]
> These two plugin tests should be enhanced to test your scenarios.
> 
> 1. https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/elasticsearch-5.x-scenario
> 2. https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/elasticsearch-6.x-scenario
> 
> If you haven't familiar with the plugin test, read this doc, https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md. You should only need to use the ES APIs in the way used to have issue, and change the expected data file to prove it is working now.

i have updated

**Comment by Indifer** [comment]
Wrong operation

**Comment by wu-sheng** [comment]
@aderm Please recheck this PR.

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    @Mock
+    private EnhancedInstance enhancedInstance;
+
+    @Mock
+    private DiscoveryNode discoveryNode;
+
+//    @Mock
+//    private SearchRequest searchRequest;
+
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: comment should delete

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        InetSocketAddress inetSocketAddress = new InetSocketAddress("122.122.122.122", 9300);
+        TransportAddress transportAddress = new TransportAddress(inetSocketAddress);
+        when(discoveryNode.getAddress()).thenReturn(transportAddress);
+
+        when(enhanceInfo.transportAddresses()).thenReturn("122.122.122.122:9300");
+        when(enhanceInfo.getClusterName()).thenReturn("skywalking-es");
+        when(enhancedInstance.getSkyWalkingDynamicField()).thenReturn(enhanceInfo);
+
+//        when(searchRequest.indices()).thenReturn(new String[]{"endpoint"});
+//        when(searchRequest.types()).thenReturn(new String[]{"searchType"});
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: also this

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+        AbstractTracingSpan getSpan = getSpan();
+        assertGetSpan(getSpan, getRequest);
+    }
+
+    @Test
+    public void testIndexRequest() throws Throwable {
+
+        AbstractTracingSpan getSpan = getSpan();
+        assertGetSpan(getSpan, getRequest);
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: `getRequest ` maybe `indexRequest`?

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Test
+    public void testUpdateRequest() throws Throwable {
+
+        AbstractTracingSpan getSpan = getSpan();
+        assertGetSpan(getSpan, getRequest);
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: `getRequest ` maybe `updateRequest`?

File: apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Test
+    public void testDeleteRequest() throws Throwable {
+
+        AbstractTracingSpan getSpan = getSpan();
+        assertGetSpan(getSpan, getRequest);
```
Comment (aderm) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-6.x-plugin/src/test/java/org/apache/skywalking/apm/plugin/elasticsearch/v6/interceptor/TransportActionNodeProxyExecuteMethodsInterceptorTest.java]: `getRequest ` maybe `deleteRequest`?

**Comment by Indifer** [comment]
@aderm I have updated

**Comment by aderm** [comment]
LGTM

**Comment by wu-sheng** [comment]
@aderm Please approve directly.

**Comment by aderm** [comment]
Done

**Pull request merged**